src/plugins/data/server/index_patterns/utils.ts

```diff
@@ -34,3 +42,152 @@ export const findIndexPatternById = async (
     return savedObjectsResponse.saved_objects[0];
   }
 };
+
+const missingFields: FieldDescriptor[] = [
+  {
+    name: '_id',
+    type: 'string',
+    searchable: true,
+    aggregatable: false,
+    readFromDocValues: false,
+    esTypes: [],
+  },
+  {
+    name: '_index',
+    type: 'string',
+    searchable: true,
+    aggregatable: true,
+    readFromDocValues: false,
+    esTypes: [],
+  },
+];
+
+/**
+ * Creates a single field item.
+ *
+ * This is a mutatious HOT CODE PATH function that will have array sizes up to 4.7 megs
+ * in size at a time calling this function repeatedly. This function should be as optimized as possible
+ * and should avoid any and all creation of new arrays, iterating over the arrays or performing
+ * any n^2 operations.
+ * @param beatFields The beat fields reference
+ * @param patternList The pattern
+ * @param fieldDescriptor The index its self
+ * @param patternListIdx The index within the patternList
+ */
+export const createFieldItem = (
+  beatFields: BeatFields,
+  patternList: string[],
+  fieldDescriptor: FieldDescriptor,
+  patternListIdx: number
+): IndexField => {
+  const alias = patternList[patternListIdx];
+  const splitFieldName = fieldDescriptor.name.split('.');
+  const fieldName =
+    splitFieldName[splitFieldName.length - 1] === 'text'
+      ? splitFieldName.slice(0, splitFieldName.length - 1).join('.')
+      : fieldDescriptor.name;
+  const beatField = beatFields[fieldName] ?? {};
+  if (isEmpty(beatField.category)) {
+    beatField.category = splitFieldName[0];
+  }
+  return {
+    ...beatField,
+    ...fieldDescriptor,
+    indexes: [alias],
+  };
+};
+
+/**
+ * This is a mutatious HOT CODE PATH function that will have array sizes up to 4.7 megs
+ * in size at a time when being called. This function should be as optimized as possible
+ * and should avoid any and all creation of new arrays, iterating over the arrays or performing
+ * any n^2 operations. The `.push`, and `forEach` operations are expected within this function
+ * to speed up performance.
+ *
+ * This intentionally waits for the next tick on the event loop to process as the large 4.7 megs
+ * has already consumed a lot of the event loop processing up to this function and we want to give
+ * I/O opportunity to occur by scheduling this on the next loop.
+ * @param beatFields The beatFields to reference
+ * @param responsesIndexFields The response index fields to loop over
+ * @param patternList The list of index patterns such as filebeat-*
+ */
+export const formatFirstFields = async (
+  beatFields: BeatFields,
+  responsesIndexFields: FieldDescriptor[][],
+  patternList: string[]
+): Promise<IndexField[]> => {
+  return new Promise((resolve) => {
+    setTimeout(() => {
```

Comment:
just wondering why `setTimeout` was added here. I see no reason to postpone the execution of this code and schedule a new macrotask for this.

it means that whatever you return from that meethod all times will be returned as async data. 

```
export const formatFirstFields = async (
  beatFields: BeatFields,
  responsesIndexFields: FieldDescriptor[][],
  patternList: string[]
): Promise<IndexField[]> => {
  return responsesIndexFields.reduce(
    (accumulator: IndexField[], indexFields: FieldDescriptor[], patternListIdx: number) => {
      [...missingFields, ...indexFields].forEach((index) => {
        const item = createFieldItem(beatFields, patternList, index, patternListIdx);
        accumulator.push(item);
      });
      return accumulator;
    },
    []
  );
};
```